### ISSUE TITLE
[Design] Tensorcery: Migrating away from libtorch

### ISSUE THREAD

**Issue by mratsim** [mratsim/tattletale#111]
This is a sketch of `Tensorcery` a new pure Nim tensor library to replace libtorch and drawing experience from:
- using PyTorch,
- writing [Arraymancer](https://github.com/mratsim/Arraymancer),
- upgrading it using [Laser backend](https://github.com/mratsim/Arraymancer/tree/master/src/arraymancer/laser)
- the ndarray julia challenge: https://nextjournal.com/sdanisch/the-julia-challenge 
  - https://github.com/mratsim/nim-julia-challenge/blob/master/src/tensor.nim
- and so many more research into tensor abstraction (Apple MLX, Google DeepMind JAX, ...)

## Scope and what this does not solve

There are several paradigms in writing tensor libraries:
- Theano (sic) and Tensorflow  "define-and-run" static graphs
- Chainer (sic) and PyTorch and Tensorflow Eager "define-by-run" dynamic graphs + eager execution
- MLX "define-by-run" + lazy execution, with ops only building a graph that is then interpreted
- JAX "define-by-run" + JIT execution, with ops building a graph that is then compiled, for example via MLIR or XLA (see #20)

i.e. a tensor library addresses "how to compose operations" and the output is either computation or computation graphs.

Another level in the stack is "how does one operation" becomes executable code. This is what Cutlass, CuteDSL, TileIR, ThunderKittens, Triton, MLIR address. In Tattletale that would be the job of the `Positron` kernel library + `Ceramic` tile algebra and layout algebra abstraction + `Crucible` Nim -> Cuda, Metal, OpenCL, Vulkan, WebGPU compiler.

For now we take a pragmatic approach that is easy to optimize and maintain. Tensorcery will hold memory or views and deal with zero-copy shape transformation (permutation, transpose, reshape) and all operations that trigger data movement or compute will be delegated to generic kernels in Positron.

## Motivations

We are an inference library, the main motivation for graphs is training and enabling reverse-mode autodifferentiation, we don't need that. For optimization, the tensor library is in my opinion the wrong level (see [PyTorch Glow](https://github.com/pytorch/glow)) especially given that optimization is hardware specific (tensor cores obviously but also see H100/B200 extremely high compute vs memory bandwidth needing TMA accelerator vs Apple extremely high bandwidth vs compute + unified memory), especially in a language that has extremely good support of generics and compile-time evaluation.

As an inference library, we focus on supporting pre-existing architectures and not random research, this means while we need to make it easy to add new custom layers (say Mamba 3) we don't need to make it easy to add a new layer for training.  There is no "fallback" pure-Nim needed, we can just use an optimized compute kernel everywhere.

### Datatypes

In Flambeau (https://github.com/SciNim/flambeau/blob/59c77fa/flambeau/tensors.nim#L14-L16) and Arraymancer/Laser (https://github.com/mratsim/Arraymancer/blob/d3b6771/src/arraymancer/laser/tensor/datatypes.nim#L19-L48)

```nim
type
  ## This works around the issue that `supportsCopyMem` is not supported in a type section
  KnownSupportsCopyMem* = concept x, type T
    supportsCopyMem(T)

  RawImmutableView*[T] = distinct ptr UncheckedArray[T]
  RawMutableView*[T] = distinct ptr UncheckedArray[T]

  Metadata* = DynamicStackArray[int]

  # On CPU, the tensor datastructures and basic accessors
  # are defined in laser/tensor/datatypes
  MetadataArray* {.deprecated: "Use Metadata instead".} = Metadata

  Tensor*[T] = object                    # Total stack: 128 bytes = 2 cache-lines
    shape*: Metadata                     # 56 bytes
    strides*: Metadata                   # 56 bytes
    offset*: int                         # 8 bytes
    storage*: CpuStorage[T]              # 8 bytes

  CpuStorage*[T] {.shallow.} = ref CpuStorageObj[T] # Total heap: 25 bytes = 1 cache-line
  CpuStorageObj[T] {.shallow.} = object
    # Workaround supportsCopyMem in type section - https://github.com/nim-lang/Nim/issues/13193
    when T is KnownSupportsCopyMem:
      raw_buffer*: ptr UncheckedArray[T]                               # 8 bytes
      memalloc*: pointer                                               # 8 bytes
      isMemOwner*: bool                                                # 1 byte
      deleter*: proc(memalloc: pointer) {.noconv, gcsafe, raises: [].} # 8 bytes
    else: # Tensors of strings, other ref types or non-trivial destructors
      raw_buffer*: seq[T]                # 8 bytes (16 for seq v2 backed by destructors?)
```

the datatype is enforced at compile-time. We cannot do that for the user-facing type, the datatype has to be runtime or it becomes really awkward to load models or even impossible for mixed-precision models.

However we do need an inner `TensorImpl[T]` type that will select the proper kernel to dispatch to for GPU kernels.

On a side note, we probably don't need to support strings for starter.

### Device placement

The Tensor or TensorImpl[T] should know where its data is. For now we won't use Nvidia Pascal+ unified memory (1000 series, 2017, https://developer.nvidia.com/blog/beyond-gpu-memory-limits-unified-memory-pascal/, https://developer.nvidia.com/blog/unified-memory-cuda-beginners/) so we have predictable latency.

I don't think there is a need or advantage for runtime device at the TensorImpl level so it can be `TensorImpl[T, Hw]` and string support is not necessary for inference.

### Allocator

Similar to Constantine, we will have dedicated per-device allocator modules https://github.com/mratsim/constantine/blob/de33a00/constantine/platforms/allocs.nim but those will be mostly for tracking allocation metrics to ease optimization, maintenance and debugging.

I don't think there is a need for a caching allocator:
- weights are per inference engine and live for a long-time (forever even unless we allow model switching) and weight allocation is dwarfed by weight loading time.
- KV cache has its own specialized PagePool allocator https://github.com/mratsim/tattletale/blob/d0d7209/workspace/transformers/src/stateful/page_pool.nim and its lifecycle is detailed here https://github.com/mratsim/tattletale/blob/d0d720945dbfa16834e7f4863ae0fddbb5cec9fb/workspace/transformers/src/stateful/orchestrator.nim#L15-L159

The allocators will do 128B alignment to allow vectorized loads everywhere.
